File: src/main/java/com/stockdownloader/model/OptionsTrade.java
package com.stockdownloader.model;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Objects;

/**
 * Tracks an individual options trade from entry to exit.
 * Each contract represents 100 shares of the underlying.
 */
public final class OptionsTrade {

    public enum Direction { BUY, SELL }
    public enum Status { OPEN, CLOSED, EXPIRED }

    private static final int CONTRACT_MULTIPLIER = 100;

    private final OptionType optionType;
    private final Direction direction;
    private final BigDecimal strike;
    private final String expirationDate;
    private final String entryDate;
    private final BigDecimal entryPremium;
    private final int contracts;
    private final long entryVolume;

    private Status status;
    private String exitDate;
    private BigDecimal exitPremium;
    private BigDecimal profitLoss;
    private BigDecimal returnPct;

    public OptionsTrade(OptionType optionType, Direction direction, BigDecimal strike,
                        String expirationDate, String entryDate, BigDecimal entryPremium,
                        int contracts, long entryVolume) {
        this.optionType = Objects.requireNonNull(optionType, "optionType must not be null");
        this.direction = Objects.requireNonNull(direction, "direction must not be null");
        this.strike = Objects.requireNonNull(strike, "strike must not be null");
        this.expirationDate = Objects.requireNonNull(expirationDate, "expirationDate must not be null");
        this.entryDate = Objects.requireNonNull(entryDate, "entryDate must not be null");
        this.entryPremium = Objects.requireNonNull(entryPremium, "entryPremium must not be null");
        if (contracts <= 0) {
            throw new IllegalArgumentException("contracts must be positive");
        }
        this.contracts = contracts;
        this.entryVolume = entryVolume;
        this.status = Status.OPEN;
        this.profitLoss = BigDecimal.ZERO;
        this.returnPct = BigDecimal.ZERO;
    }

    /**
     * Close the trade at the given premium.
     */
    public void close(String exitDate, BigDecimal exitPremium) {
        Objects.requireNonNull(exitDate, "exitDate must not be null");
        Objects.requireNonNull(exitPremium, "exitPremium must not be null");
        if (this.status != Status.OPEN) {
            throw new IllegalStateException("Trade is not open, current status: " + status);
        }
        this.exitDate = exitDate;
        this.exitPremium = exitPremium;
        this.status = Status.CLOSED;
        calculateProfitLoss();
    }

    /**
     * Mark the trade as expired (option expired worthless or exercised).
     */
    public void expire(String expiryDate, BigDecimal settlementPremium) {
        Objects.requireNonNull(expiryDate, "expiryDate must not be null");
        Objects.requireNonNull(settlementPremium, "settlementPremium must not be null");
        if (this.status != Status.OPEN) {
            throw new IllegalStateException("Trade is not open, current status: " + status);
        }
        this.exitDate = expiryDate;
        this.exitPremium = settlementPremium;
        this.status = Status.EXPIRED;
        calculateProfitLoss();
    }

    private void calculateProfitLoss() {
        BigDecimal premiumDiff = switch (direction) {
            case BUY -> exitPremium.subtract(entryPremium);
            case SELL -> entryPremium.subtract(exitPremium);
        };
        this.profitLoss = premiumDiff
                .multiply(BigDecimal.valueOf(contracts))
                .multiply(BigDecimal.valueOf(CONTRACT_MULTIPLIER));

        BigDecimal totalCost = entryPremium
                .multiply(BigDecimal.valueOf(contracts))
                .multiply(BigDecimal.valueOf(CONTRACT_MULTIPLIER));
        if (totalCost.compareTo(BigDecimal.ZERO) != 0) {
            this.returnPct = profitLoss
                    .divide(totalCost, 6, RoundingMode.HALF_UP)
                    .multiply(BigDecimal.valueOf(100));
        }
    }

    public boolean isWin() {
        return profitLoss.compareTo(BigDecimal.ZERO) > 0;
    }

    /**
     * Total premium paid/received at entry (contracts * 100 * premium).
     */
    public BigDecimal totalEntryCost() {
        return entryPremium.multiply(BigDecimal.valueOf((long) contracts * CONTRACT_MULTIPLIER));
    }

    public OptionType getOptionType() { return optionType; }
    public Direction getDirection() { return direction; }
    public BigDecimal getStrike() { return strike; }
    public String getExpirationDate() { return expirationDate; }
    public String getEntryDate() { return entryDate; }
    public BigDecimal getEntryPremium() { return entryPremium; }
    public int getContracts() { return contracts; }
    public long getEntryVolume() { return entryVolume; }
    public Status getStatus() { return status; }
    public String getExitDate() { return exitDate; }
    public BigDecimal getExitPremium() { return exitPremium; }
    public BigDecimal getProfitLoss() { return profitLoss; }
    public BigDecimal getReturnPct() { return returnPct; }

    @Override
    public String toString() {
        return "%s %s %s $%s exp:%s | Entry:%s @$%s -> Exit:%s @$%s | P/L:$%s (%.2f%%) vol:%d".formatted(
                direction, optionType, status, strike.setScale(2, RoundingMode.HALF_UP),
                expirationDate,
                entryDate, entryPremium.setScale(2, RoundingMode.HALF_UP),
                exitDate != null ? exitDate : "N/A",
                exitPremium != null ? exitPremium.setScale(2, RoundingMode.HALF_UP) : "N/A",
                profitLoss.setScale(2, RoundingMode.HALF_UP),
                returnPct.doubleValue(),
                entryVolume);
    }
}

File: stockdownloader/model/options_trade.py
from enum import Enum, auto

from .option_type import OptionType


class OptionsDirection(Enum):
    BUY = auto()
    SELL = auto()


class OptionsTradeStatus(Enum):
    OPEN = auto()
    CLOSED = auto()
    EXPIRED = auto()


class OptionsTrade:
    """Tracks an individual options trade from entry to exit.
    Each contract represents 100 shares of the underlying.
    """

    CONTRACT_MULTIPLIER: int = 100

    def __init__(
        self,
        option_type: OptionType,
        direction: OptionsDirection,
        strike: float,
        expiration_date: str,
        entry_date: str,
        entry_premium: float,
        contracts: int,
        entry_volume: int,
    ):
        if contracts <= 0:
            raise ValueError("contracts must be positive")
        self.option_type = option_type
        self.direction = direction
        self.strike = strike
        self.expiration_date = expiration_date
        self.entry_date = entry_date
        self.entry_premium = entry_premium
        self.contracts = contracts
        self.entry_volume = entry_volume
        self.status = OptionsTradeStatus.OPEN
        self.exit_date: str | None = None
        self.exit_premium: float | None = None
        self.profit_loss: float = 0.0
        self.return_pct: float = 0.0

    def close(self, exit_date: str, exit_premium: float) -> None:
        """Close the trade at the given premium."""
        if self.status != OptionsTradeStatus.OPEN:
            raise RuntimeError(f"Trade is not open, current status: {self.status.name}")
        self.exit_date = exit_date
        self.exit_premium = exit_premium
        self.status = OptionsTradeStatus.CLOSED
        self._calculate_profit_loss()

    def expire(self, expiry_date: str, settlement_premium: float) -> None:
        """Mark the trade as expired (option expired worthless or exercised)."""
        if self.status != OptionsTradeStatus.OPEN:
            raise RuntimeError(f"Trade is not open, current status: {self.status.name}")
        self.exit_date = expiry_date
        self.exit_premium = settlement_premium
        self.status = OptionsTradeStatus.EXPIRED
        self._calculate_profit_loss()

    def _calculate_profit_loss(self) -> None:
        assert self.exit_premium is not None
        if self.direction == OptionsDirection.BUY:
            premium_diff = self.exit_premium - self.entry_premium
        else:
            premium_diff = self.entry_premium - self.exit_premium

        self.profit_loss = premium_diff * self.contracts * self.CONTRACT_MULTIPLIER

        total_cost = self.entry_premium * self.contracts * self.CONTRACT_MULTIPLIER
        if total_cost != 0:
            self.return_pct = (self.profit_loss / total_cost) * 100
        else:
            self.return_pct = 0.0

    @property
    def is_win(self) -> bool:
        return self.profit_loss > 0

    def total_entry_cost(self) -> float:
        """Total premium paid/received at entry (contracts * 100 * premium)."""
        return self.entry_premium * self.contracts * self.CONTRACT_MULTIPLIER

    def __repr__(self) -> str:
        exit_d = self.exit_date or "N/A"
        exit_p = f"${self.exit_premium:.2f}" if self.exit_premium is not None else "N/A"
        return (
            f"{self.direction.name} {self.option_type.name} {self.status.name} "
            f"${self.strike:.2f} exp:{self.expiration_date} | "
            f"Entry:{self.entry_date} @${self.entry_premium:.2f} -> "
            f"Exit:{exit_d} @{exit_p} | P/L:${self.profit_loss:.2f} "
            f"({self.return_pct:.2f}%) vol:{self.entry_volume}"
        )
